aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,376 @@
+
+        // get the actual committed balance
+        let actual_comm = pedersen::new_commitment_from_compressed(&private_balance<CoinType>(owner));
+
+        pedersen::commitment_equals(&actual_comm, &expected_comm)
+    }
+
+    /// Initializes a veiled coin store for the specified account.
+    public fun register<CoinType>(account: &signer) {
+        let account_addr = signer::address_of(account);
+        assert!(
+            !has_veiled_coin_store<CoinType>(account_addr),
+            error::already_exists(EVEILED_COIN_STORE_ALREADY_PUBLISHED),
+        );
+
+        let coin_store = VeiledCoinStore<CoinType> {
+            private_balance: ristretto255::point_identity_compressed(),
+            deposit_events: account::new_event_handle<DepositEvent>(account),
+            withdraw_events: account::new_event_handle<WithdrawEvent>(account),
+        };
+        move_to(account, coin_store);
+    }
+
+    /// Mints a veiled coin from a normal coin, shelving the normal coin into the resource account's coin store.
+    ///
+    /// WARNING: Fundamentally, there is no way to hide the value of the coin being minted here.
+    public fun mint_from_coin<CoinType>(c: Coin<CoinType>): VeiledCoin<CoinType> acquires VeiledCoinMinter {
+        // If there is no CoinStore<CoinType> yet, create one.
+        let rsrc_acc_signer = get_resource_account_signer();
+        let rsrc_acc_addr = signer::address_of(&rsrc_acc_signer);
+        if(!coin::is_account_registered<CoinType>(rsrc_acc_addr)) {
+            coin::register<CoinType>(&rsrc_acc_signer);
+        };
+
+        // Move the traditional coin into the coin store, so we can mint a veiled coin.
+        // (There is no other way to drop a traditional coin, for safety reasons, so moving it into a coin store is
+        //  the only option.)
+        let value = new_scalar_from_u64(coin::value(&c));
+        coin::deposit(rsrc_acc_addr, c);
+
+        VeiledCoin<CoinType> {
+            private_value: pedersen::new_non_hiding_commitment_for_bulletproof(&value)
+        }
+    }
+
+    /// Sends the specified private amount to `recipient` and updates the private balance of `sender`. Requires a range
+    /// proof on the new balance of the sender, to ensure the sender has enough money to send. `private_amount` is a
+    /// Pedersen commitment to the amount being sent.
+    public fun transfer_privately_to<CoinType>(
+        sender: &signer,
+        recipient: address,
+        private_amount: Commitment,
+        range_proof: &RangeProof)
+    acquires VeiledCoinStore {
+        let vc = withdraw<CoinType>(sender, private_amount, range_proof);
+
+        deposit(recipient, vc);
+    }
+
+    /// Sends the specified public amount to `recipient` and updates the private balance of `sender`. Requires a range
+    /// proof on the new balance of the sender, to ensure the sender has enough money to send.
+    public fun transfer_publicly_to<CoinType>(
```

Comment:
Is there an example use case of public veiled coin transfers? Wondering if this could erode the privacy guarantees. I.e. say someone receives all their VeiledCoin privately. Then someone makes a public transaction of N coins without the recipient's consent to that address. Then everyone knows that address has at least N coins where previously there was no public information about it.

Reply:
Let's remove this for now. Not sure what use case I had in mind.